Codeinwp/themeisle-sdk
# feat: allow upsell notification for new users

### Summary
Added a new module called Welcome
The module will use a filter `<product_slug>_welcome_metadata` to display an upsell notification for users that have used the product for more than 7 days, the notification is only displayed for 5 days ( after 12 days since the product installation it will not be registered anymore. )

The products that require a upsell message can use the following filter:
```php
add_filter( '<product_slug>_welcome_metadata', function() {
     return [
        'is_enabled' => <condition_if_pro_available>,
        'cta_link' => tsdk_utmify( 'https://link_to_upgrade.with/?discount=<discountCode>')
     ];
} );
```

Changed the Notification module to allow for an optional image and control over the notification type.
When using an image the Image and message will be wrapped by a flex box.

Included PHPUnit test for the new module.

### Screenshots
<img width="2384" alt="image" src="https://user-images.githubusercontent.com/23024731/234552172-d287e67f-16ae-49db-871f-70f1932473a9.png">

<img width="2372" alt="image" src="https://user-images.githubusercontent.com/23024731/234552245-75ee1894-d1ef-4b5b-86cf-f8867c1cce2e.png">


### How to test

1. On a fresh instance
2. Use the current version of the SDK
3. For Neve make sure Neve Pro is not active / For Otter make sure the Pro License is not enabled.
4. Add the following filters:
    ```php
    add_filter('neve_welcome_metadata', function(){
        return [
            'is_enabled' => !defined('NEVE_PRO_VERSION'),
            'pro_name' => 'Neve Pro Addon',
            'logo' => get_template_directory_uri() . '/assets/img/dashboard/logo.svg',
            'cta_link' => tsdk_utmify( 'https://themeisle.com/themes/neve/pricing/?discount=newpirate10', 'neve-welcome', 'notice')
        ];
    });

    add_filter('otter_blocks_welcome_metadata', function(){
        return [
            'is_enabled' => ! Pro::is_pro_active(),
           'pro_name' => 'Otter PRO',
            'logo' => OTTER_BLOCKS_URL . 'assets/images/logo-alt.png',
            'cta_link' => tsdk_utmify( 'https://themeisle.com/plugins/otter-blocks/pricing/?discount=newpirate10', 'otter-blocks-welcome', 'notice')
        ];
    });
   ```
5. Check that even though the filters are added the Notice is not displayed
6. Update the install time for each product to 7 days earlier:
    ```bash
    wp option update neve_install <unix_time_in_seconds>
    wp option update otter_blocks_install  <unix_time_in_seconds>
    ```
7. Check that the notice is displayed and the link to upgrade goes to the link from the filter
8. Check that on refresh if both filters are added they are randomly displayed
9. Check no other welcome Notifications are displayed E.g. Asking for Review
10. Change the install time using the commands from 6. to 13 days earlier
11. Check that the notification is not displayed anymore.
12. Change the install time using the commands from 6. to 6 days earlier
13. Check that the notification is not displayed.
14. Change back the time to 7 days earlier and dismiss the notification. Check that is not displayed after dismissal.

Closes: Codeinwp/themeisle#1472.
